README.md
<img src="https://raw.githubusercontent.com/js-data/js-data/master/js-data.png" alt="js-data logo" title="js-data" align="right" width="96" height="96" />

# js-data-rethinkdb

[![Slack][1]][2]
[![NPM][3]][4]
[![Tests][5]][6]
[![Downloads][7]][8]
[![Coverage][9]][10]

RethinkDB adapter for the [JSData Node.js ORM][11].

### Installation

    npm install --save js-data js-data-rethinkdb

### Usage

```js
import { RethinkDBAdapter } from 'js-data-rethinkdb';

// Create an instance of RethinkDBAdapter
const adapter = new RethinkDBAdapter();

// Other JSData setup hidden

// Register the adapter instance
store.registerAdapter('rethinkdb', adapter, { default: true });
```

### JSData + RethinkDB Tutorial

Start with the [JSData + RethinkDB tutorial][12] or checkout the [API Reference Documentation][13].

### Need help?

Please [post a question][14] on Stack Overflow. **This is the preferred method.**

You can also chat with folks on the [Slack Channel][15]. If you end up getting
your question answered, please still consider consider posting your question to
Stack Overflow (then possibly answering it yourself). Thanks!

### Want to contribute?

Awesome! You can get started over at the [Contributing guide][16].

Thank you!

### License

[The MIT License (MIT)][17]

Copyright (c) 2014-2017 [js-data-rethinkdb project authors][18]

[1]: http://slack.js-data.io/badge.svg
[2]: http://slack.js-data.io
[3]: https://img.shields.io/npm/v/js-data-rethinkdb.svg?style=flat
[4]: https://www.npmjs.org/package/js-data-rethinkdb
[5]: https://img.shields.io/circleci/project/js-data/js-data-rethinkdb.svg?style=flat
[6]: https://circleci.com/gh/js-data/js-data-rethinkdb
[7]: https://img.shields.io/npm/dm/js-data-rethinkdb.svg?style=flat
[8]: https://www.npmjs.org/package/js-data-rethinkdb
[9]: https://img.shields.io/codecov/c/github/js-data/js-data-rethinkdb.svg?style=flat
[10]: https://codecov.io/github/js-data/js-data-rethinkdb
[11]: http://www.js-data.io/
[12]: http://www.js-data.io/docs/js-data-rethinkdb
[13]: http://api.js-data.io/js-data-rethinkdb
[14]: http://stackoverflow.com/questions/tagged/jsdata
[15]: http://slack.js-data.io/
[16]: https://github.com/js-data/js-data-rethinkdb/blob/master/.github/CONTRIBUTING.md
[17]: https://github.com/js-data/js-data-rethinkdb/blob/master/LICENSE
[18]: https://github.com/js-data/js-data-rethinkdb/blob/master/AUTHORS


CHANGELOG.md
##### 3.0.0 - 18 August 2017

Stable 3.0.0 release

##### 3.0.0-rc.3 - 17 October 2016

###### Bug fixes
- #24 - Fixed update throwing error when record was unchanged.

##### 3.0.0-rc.2 - 23 August 2016

###### Backwards compatible changes
- Add support for `fields` option to `find` and `findAll`.

##### 3.0.0-rc.1 - 15 August 2016

###### Backwards compatible changes
- Upgrade dependencies

##### 3.0.0-beta.8 - 08 July 2016

###### Backwards compatible changes
- RethinkDBAdapter now inherits from Adapter via the `extend` method

##### 3.0.0-beta.7 - 29 June 2016

###### Backwards compatible changes
- Upgraded dependencies

##### 3.0.0-beta.6 - 17 May 2016

###### Backwards compatible changes
- Added support for grouped "where" clauses

##### 3.0.0-beta.5 - 10 May 2016

###### Breaking changes
- Removed default import, which wasn't working

##### 3.0.0-beta.4 - 10 May 2016

###### Breaking changes
- Moved `rethinkdbdash` options to `RethinkDBAdapter#rOpts`.

##### 3.0.0-beta.3 - 30 April 2016

- Upgraded to `js-data-adapter` v0.6.1

##### 3.0.0-beta.2 - 29 April 2016

###### Breaking changes
- How you must now import in ES2015:

    ```js
    import RethinkDBAdapter from 'js-data-rethinkdb'
    const adapter = new RethinkDBAdapter()
    ```
    or
    ```js
    import {RethinkDBAdapter, version} from 'js-data-rethinkdb'
    console.log(version)
    const adapter = new RethinkDBAdapter()
    ```

- How you must now import in ES5:

    ```js
    var JSDataRethinkDB = require('js-data-rethinkdb')
    var RethinkDBAdapter = JSDataRethinkDB.RethinkDBAdapter
    var adapter = new RethinkDBAdapter()
    ```

- `OPERATORS` is now an export, not a static property of the `RethinkDBAdapter` class.
- Moved some `dist` files to `release` to reduce noise

###### Other
- Upgraded dependencies
- Improved JSDoc comments
- Now using js-data JSDoc template

##### 3.0.0-beta.1 - 17 April 2016

Official v3 beta release

###### Other
- Upgraded dependencies

##### 3.0.0-alpha.13 - 17 March 2016

###### Backwards incompatible API changes
- Added sum and count methods

##### 3.0.0-alpha.12 - 10 March 2016

###### Other
- Moved more common adapter functionality into js-data-adapter

##### 3.0.0-alpha.11 - 06 March 2016

###### Backwards compatible bug fixes
- create() now makes use of withoutRelations()
- More use of Adapter#respond() more consistent

##### 3.0.0-alpha.10 - 06 March 2016

###### Backwards compatible API changes
- Now making use of makeHasManyForeignKeys

##### 3.0.0-alpha.9 - 06 March 2016

###### Other
- Upgraded js-data-adapter

##### 3.0.0-alpha.8 - 06 March 2016

###### Other
- Extracted common adapter functionality into js-data-adapter
- Extracted common js-data repo utility scripts into js-data-repo-tools

##### 3.0.0-alpha.7 - 03 March 2016

###### Backwards compatible API changes
- Added RethinkDBAdapter.extend

##### 3.0.0-alpha.6 - 01 March 2016

###### Backwards compatible bug fixes
- Couple of fixes based on the latest js-data-adapter-tests

##### 3.0.0-alpha.5 - 27 February 2016

###### Backwards compatible API changes
- Added ability to override query operators used by RethinkDBAdapter#filterSequence.

###### Other
- Improved JSDoc comments.

##### 3.0.0-alpha.4 - 26 February 2016

###### Other
- Finished JSDoc comments

##### 3.0.0-alpha.3 - 26 February 2016

###### Backwards compatible API changes
- Removed returnDeletedIds option
- Added insertOpts, updateOpts, deleteOpts, and runOpts options

##### 3.0.0-alpha.2 - 26 February 2016

###### Backwards compatible bug fixes
- Fixed edge case with returnDeletedIds

##### 3.0.0-alpha.1 - 26 February 2016

###### Breaking API changes
- Now depends on js-data 3.x
- Now longer uses internal `defaults` property, settings are on the adapter instance itself

###### Backwards compatible API changes
- Added createMany and updateMany methods
- Added lifecycle hooks for all methods
- Added contains, notContains, |contains, and |notContains operators

##### 2.2.0 - 25 February 2016

###### Other
- General improvements

##### 2.1.0 - 26 October 2015

###### Backwards compatible bug fixes
- #16 - Handle cursor errors by @BBB

###### Other
- Removed Grunt
- Tests now use yield and generators

##### 2.0.3 - 13 October 2015

###### Backwards compatible bug fixes
- #13, #14 - Using `where` in a query without an operator breaks by @internalfx

##### 2.0.2 - 09 September 2015

###### Backwards compatible bug fixes
- #9 - CI tests failing because of outdated npm

##### 2.0.1 - 03 July 2015

###### Backwards compatible bug fixes
- #8 - update/updateAll fail if there are no changes

##### 2.0.0 - 02 July 2015

Stable Version 2.0.0

##### 1.2.0 - 26 March 2015

###### Backwards compatible API changes
- #3 - Add support for loading relations
- #4 - Add support for auto-creating secondary indexes based on the relation definitions

###### Backwards compatible bug fixes
- #5 - Should not be saving relations (duplicating data)
- #6 - Need to use removeCircular

##### 1.1.2 - 07 March 2015

###### Other
- Converted to ES6, using Babel.js to transpile to ES5.

##### 1.1.1 - 25 February 2015

- Updated dependencies

##### 1.1.0 - 05 February 2015

- #3, #4 - Added the `with` option to `DSRethinkDBAdapter#find` which allows for loading relations.

##### 1.0.0 - 03 February 2015

Stable Version 1.0.0

##### 1.0.0-beta.1 - 12 January 2015

Now in beta.

##### 1.0.0-alpha.5 - 28 November 2014

###### Backwards compatible API changes
- Added isectEmpty, isectNotEmpty, |isectEmpty, and |isectNotEmpty filter operators.

##### 1.0.0-alpha.4 - 23 November 2014

###### Backwards compatible bug fixes
- Fixed improper use of r.row in nested queries

##### 1.0.0-alpha.3 - 23 November 2014

###### Backwards compatible bug fixes
- Fixed the "in", "notIn", "|in", and "|notIn" operators

##### 1.0.0-alpha.2 - 23 November 2014

Upgraded dependencies

##### 1.0.0-alpha.1 - 01 November 2014

###### Backwards compatible API changes
- #1 - auto create db and tables if they don't exist


test/find.spec.js
var Promise = require('bluebird')
describe('DSRethinkDBAdapter#find', function() {
  it('should find a user in RethinkDB', function*() {
    var user = yield adapter.create(User, {
      name: 'John'
    })
    assert.equal(user.name, 'John')
    assert.isString(user.id)
    user = yield adapter.find(User, user.id)
    assert.equal(user.name, 'John')
    assert.isString(user.id)
    assert.deepEqual(user, {
      id: user.id,
      name: 'John'
    })
    var post = yield adapter.create(Post, {
      content: 'test',
      userId: user.id
    })
    assert.equal(post.content, 'test')
    assert.isString(post.id)
    assert.isString(post.userId)
    var comments = yield Promise.all([
      adapter.create(Comment, {
        content: 'test2',
        postId: post.id,
        userId: user.id
      }),
      adapter.create(Comment, {
        content: 'test3',
        postId: post.id,
        userId: user.id
      })
    ])
    comments.sort(function(a, b) {
      return a.content > b.content
    })
    post = yield adapter.find(Post, post.id, {
      with: ['user', 'comment']
    })
    post.comments.sort(function(a, b) {
      return a.content > b.content
    })
    assert.equal(JSON.stringify(post.user), JSON.stringify(user))
    assert.equal(JSON.stringify(post.comments), JSON.stringify(comments))
    yield adapter.destroy(User, user.id)
    try {
      yield adapter.find(User, user.id)
      throw new Error('Should not have reached here!')
    }
    catch (err) {
      assert.equal(err.message, 'Not Found!')
    }

    yield adapter.destroy(Post, post.id)
  })
})


test/create.spec.js
describe('DSRethinkDBAdapter#create', function() {
  it('should create a user in RethinkDB', function*() {
    var id
    var user = yield adapter.create(User, {
      name: 'John'
    })
    id = user.id
    assert.equal(user.name, 'John')
    assert.isString(user.id)
    user = yield adapter.find(User, user.id)
    assert.equal(user.name, 'John')
    assert.isString(user.id)
    assert.deepEqual(user, {
      id: id,
      name: 'John'
    })
    user = yield adapter.destroy(User, user.id)
    assert.isFalse(!!user)
    try {
      user = yield adapter.find(User, id)
      throw new Error('Should not have reached here!')
    }
    catch (err) {
      assert.equal(err.message, 'Not Found!')
    }
  })
})


test/update.spec.js
describe('DSRethinkDBAdapter#update', function() {
  it('should update a user in RethinkDB', function*() {
    var id
    var user = yield adapter.create(User, {
      name: 'John'
    })
    assert.equal(user.name, 'John')
    assert.isString(user.id)
    user = yield adapter.find(User, user.id)
    assert.equal(user.name, 'John')
    assert.isString(user.id)
    assert.deepEqual(user, {
      id: user.id,
      name: 'John'
    })
    user = yield adapter.update(User, user.id, {
      name: 'Johnny'
    })
    assert.equal(user.name, 'Johnny')
    assert.isString(user.id)
    assert.deepEqual(user, {
      id: user.id,
      name: 'Johnny'
    })
    user = yield adapter.find(User, user.id)
    assert.equal(user.name, 'Johnny')
    assert.isString(user.id)
    assert.deepEqual(user, {
      id: user.id,
      name: 'Johnny'
    })
    user = yield adapter.destroy(User, user.id)
    assert.isFalse(!!user)
  })
  it('should still work if there are no changes', function*() {
    var user = yield adapter.create(User, {
      name: 'John'
    })
    assert.equal(user.name, 'John')
    assert.isString(user.id)
    user = yield adapter.find(User, user.id)
    assert.equal(user.name, 'John')
    assert.isString(user.id)
    assert.deepEqual(user, {
      id: user.id,
      name: 'John'
    })
    user = yield adapter.update(User, user.id, user)
    assert.equal(user.name, 'John')
    assert.isString(user.id)
    assert.deepEqual(user, {
      id: user.id,
      name: 'John'
    })
    user = yield adapter.find(User, user.id)
    assert.equal(user.name, 'John')
    assert.isString(user.id)
    assert.deepEqual(user, {
      id: user.id,
      name: 'John'
    })
    user = yield adapter.destroy(User, user.id)
    assert.isFalse(!!user)
  })
})


Gruntfile.js
/*
 * js-data-rethinkdb
 * https://github.com/js-data/js-data-rethinkdb
 *
 * Copyright (c) 2014-2015 Jason Dobry <http://www.js-data.io/docs/dsrethinkdbadapter>
 * Licensed under the MIT license. <https://github.com/js-data/js-data-rethinkdb/blob/master/LICENSE>
 */
module.exports = function (grunt) {
  'use strict';

  require('jit-grunt')(grunt, {
    coveralls: 'grunt-karma-coveralls'
  });
  require('time-grunt')(grunt);

  var pkg = grunt.file.readJSON('package.json');

  // Project configuration.
  grunt.initConfig({
    pkg: pkg,
    coveralls: {
      options: {
        coverage_dir: 'coverage'
      }
    },
    mochaTest: {
      all: {
        options: {
          timeout: 20000,
          reporter: 'spec'
        },
        src: ['mocha.start.js', 'test/**/*.js']
      }
    },
    webpack: {
      dist: {
        debug: true,
        entry: './src/index.js',
        output: {
          filename: './dist/js-data-rethinkdb.js',
          libraryTarget: 'commonjs2',
          library: 'js-data-rethinkdb'
        },
        externals: [
          'mout/string/underscore',
          'bluebird',
          'js-data',
          'rethinkdbdash'
        ],
        module: {
          loaders: [
            { test: /(src)(.+)\.js$/, exclude: /node_modules/, loader: 'babel-loader?blacklist=useStrict' }
          ],
          preLoaders: [
            {
              test: /(src)(.+)\.js$|(test)(.+)\.js$/, // include .js files
              exclude: /node_modules/, // exclude any and all files in the node_modules folder
              loader: "jshint-loader?failOnHint=true"
            }
          ]
        }
      }
    }
  });

  grunt.registerTask('standard', function () {
    var child_process = require('child_process');
    var done = this.async();
    grunt.log.writeln('Linting for correcting formatting...');
    child_process.exec('node node_modules/standard/bin/cmd.js --parser babel-eslint src/index.js', function (err, stdout) {
      console.log(stdout);
      if (err) {
        grunt.log.writeln('Failed due to ' + (stdout.split('\n').length - 2) + ' lint errors!');
        done(err);
      } else {
        grunt.log.writeln('Done linting.');
        done();
      }
    });
  });

  grunt.registerTask('build', [
    'standard',
    'webpack'
  ]);
  grunt.registerTask('test', ['build', 'mochaTest']);
  grunt.registerTask('default', ['build']);
};


mocha.start.js
'use strict'

// prepare environment for js-data-adapter-tests
import 'babel-polyfill'

import * as JSData from 'js-data'
import JSDataAdapterTests from './node_modules/js-data-adapter/dist/js-data-adapter-tests'
import * as JSDataRethinkDB from './src/index'

const assert = global.assert = JSDataAdapterTests.assert
global.sinon = JSDataAdapterTests.sinon

JSDataAdapterTests.init({
  debug: false,
  JSData: JSData,
  Adapter: JSDataRethinkDB.RethinkDBAdapter,
  adapterConfig: {
    rOpts: {
      buffer: 1,
      max: 5
    }
  },
  // js-data-rethinkdb does NOT support these features
  xfeatures: [
    'findAllLikeOp',
    'filterOnRelations'
  ]
})

describe('RethinkDBAdapter#handleErrors(err)', function () {
  it('should do nothing when passed a falsy value', function () {
    var Test = this
    assert.doesNotThrow(function () {
      Test.$$adapter._handleErrors(false)
    })
  })
  it('should do nothing when errors is 0', function () {
    var Test = this
    assert.doesNotThrow(function () {
      Test.$$adapter._handleErrors({
        errors: 0
      })
    })
  })
  it('should throw an error when errors > 0 && first_error is a string', function () {
    var Test = this
    var errorString = 'error string'
    assert.throws(function () {
      Test.$$adapter._handleErrors({
        errors: 1,
        first_error: errorString
      })
    }, Error, errorString)
  })
  it('should throw a generic error when errors > 0 && first_error is nothing', function () {
    var Test = this
    assert.throws(function () {
      Test.$$adapter._handleErrors({
        errors: 1
      })
    }, Error, 'Unknown RethinkDB Error')
  })
})

describe('exports', function () {
  it('should have correct exports', function () {
    assert(JSDataRethinkDB.RethinkDBAdapter)
    assert(JSDataRethinkDB.OPERATORS)
    assert(JSDataRethinkDB.OPERATORS['=='])
    assert(JSDataRethinkDB.version)
  })
})

describe('RethinkDBAdapter#find', function () {
  it('should allow subset of fields to be returned', function () {
    const User = this.$$User
    let props = { name: 'John', age: 30, role: 'admin', password: 'foo', address: { state: 'TX', zip: 12345 } }
    let user
    assert.debug('create', User.name, props)
    return this.$$adapter.create(User, props)
      .then((_user) => {
        user = _user

        return this.$$adapter.find(User, user[User.idAttribute], { fields: User.idAttribute })
      })
      .then((_user) => {
        assert.objectsEqual(_user, {
          [User.idAttribute]: user[User.idAttribute]
        }, 'only ID was selected')

        return this.$$adapter.find(User, user[User.idAttribute], { fields: [User.idAttribute, 'age'] })
      })
      .then((_user) => {
        assert.objectsEqual(_user, {
          [User.idAttribute]: user[User.idAttribute],
          age: user.age
        }, 'only ID and age were selected')
      })
  })
})

describe('RethinkDBAdapter#findAll', function () {
  it('should allow subset of fields to be returned', function () {
    const User = this.$$User
    let props = { name: 'John', age: 30, role: 'dev', password: 'foo', address: { state: 'TX', zip: 12345 } }
    let props2 = { name: 'Sally', age: 28, role: 'admin', password: 'bar', address: { state: 'MA', zip: 54321 } }
    let users
    assert.debug('create', User.name, [props, props2])
    return this.$$adapter.createMany(User, [props, props2])
      .then((_users) => {
        users = _users

        return this.$$adapter.findAll(User, { orderBy: 'name' }, { fields: User.idAttribute })
      })
      .then((_users) => {
        assert.objectsEqual(_users, [{
          [User.idAttribute]: users[0][User.idAttribute]
        }, {
          [User.idAttribute]: users[1][User.idAttribute]
        }], 'only ID was selected')

        return this.$$adapter.findAll(User, { orderBy: 'name' }, { fields: [User.idAttribute, 'age'] })
      })
      .then((_users) => {
        assert.objectsEqual(_users, [{
          [User.idAttribute]: users[0][User.idAttribute],
          age: users[0].age
        }, {
          [User.idAttribute]: users[1][User.idAttribute],
          age: users[1].age
        }], 'only ID and age were selected')
      })
  })
})
